### PR TITLE
Refactor 2fa flow

### PR DESCRIPTION
**Pull request by JC-Coder**
# Pull Request Template

## Description

Refactor flow for 2fa authentication

## Related Issue

- Fixes # (issue number)

## Contribution Guidelines

Before submitting this pull request, please review our [Contribution Guidelines](https://github.com/abeg-help/backend/blob/dev/CONTRIBUTING.md) to understand how to contribute to this project.

## Checklist

- [ ] I have reviewed the Contribution Guidelines linked above.
- [ ] I have tested my changes thoroughly and ensured that all existing tests pass.
- [ ] I have provided clear and concise commit messages.
- [ ] I have updated the project's documentation as necessary.

### CODE REVIEW

File: src/controllers/auth/signin.ts [hunk 1]
```diff
@@ -15,7 +15,7 @@ export const signIn = catchAsync(async (req: Request, res: Response) => {
 	}
 
 	const user = await User.findOne({ email, provider: Provider.Local }).select(
-		'+refreshToken +loginRetries +isSuspended +isEmailVerified +lastLogin +password'
+		'+refreshToken +loginRetries +isSuspended +isEmailVerified +lastLogin +password +timeBased2FA'
```
Comment (1finedev) [inline src/controllers/auth/signin.ts:18]: `+timeBased2FA` does not exist anymore in the user model

File: src/controllers/auth/verify2fa.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import { UserModel } from '../../models';
+import AppError from '../../common/utils/appError';
+import { twoFactorTypeEnum } from '../../common/constants';
+import { DateTime } from 'luxon';
+
+export const verifyTimeBased2fa = catchAsync(async (req: Request, res: Response) => {
+	const { user } = req;
+	const { token } = req.body;
+
+	const userFromDb = await UserModel.findOne({ email: user?.email }).select(
```
Comment (1finedev) [inline src/controllers/auth/verify2fa.ts:13]: `+timeBased2FA` does not exist anymore in the user model

File: src/routes/authRouter.ts [hunk 3]
```diff
@@ -23,13 +24,14 @@ router.post('/password/forgot', forgotPassword);
 router.post('/password/reset', resetPassword);
 router.post('/verify-email', verifyEmail);
 router.post('/resend-verification', resendVerification);
-router.post('/2fa/code/email', get2faCodeViaEmail);
```
Comment (1finedev) [inline src/routes/authRouter.ts:26]:
```
router.post('/2fa/code/email', get2faCodeViaEmail);
router.post('/2fa/time/verify', verifyTimeBased2fa);
```

These should be protected enpoints

File: src/controllers/auth/complete2faSetup.ts [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+	if (!user) {
+		throw new AppError('Unauthorized, kindly login again.');
+	}
+
+	if (user?.twoFA.active) {
+		throw new AppError('2FA is already active', 400);
+	}
+
+	if (twoFactorType === twoFactorTypeEnum.APP) {
+		const decryptedSecret = await decodeData(user.twoFA.secret as string);
```
Comment (1finedev) [inline src/controllers/auth/complete2faSetup.ts:36]: `user.twoFA.secret` will not be here as it is select false on the user model!

**Review by 1finedev — CHANGES_REQUESTED** [review]
@Hydradaboss @JC-Coder 

Kindly check comments

File: src/controllers/auth/complete2faSetup.ts (hunk 4 above)
Comment (JC-Coder) [reply src/controllers/auth/complete2faSetup.ts:36]: It is coming from the cache so it'll be available

File: src/controllers/auth/get2faCodeViaEmail.ts [hunk 5]
```diff
@@ -1,41 +1,21 @@
...
 
 export const get2faCodeViaEmail = catchAsync(async (req: Request, res: Response) => {
-	const { email } = req.body;
-
-	if (!email) {
-		throw new AppError('Email is required', 400);
-	}
-
-	const user = await UserModel.findOne({ email });
+	const { user } = req;
```
[the remark this answers is not in the document]
Comment (JC-Coder) [reply src/controllers/auth/get2faCodeViaEmail.ts:8]: it's below the protected endpoint

File: src/routes/authRouter.ts (hunk 3 above)
Comment (JC-Coder) [reply src/routes/authRouter.ts:26]: they are protected

**Pull request merged**